### PR TITLE
New GH-Pages

### PR DESCRIPTION
**Pull request by hughrawlinson**
A huge spate of changes, bear with me.

* A whole new display site.
* Using Three.js to implement a loose clone of Ge Wang's `sndpeek` in Javascript with WebGL
  * It performs much better now, and is responsive
* Modularised frontend code
* Links to relevant resources

Future ideas:
* Either
  * Remove the dependency on Browserify
  * Gulpify things (so that building and deploying the site isn't such a performance)
* Display Loudness and other features in sndpeek
* Account for logarithmic spectrum
* Grab documentation from the wiki on build, and then use JS to update on page load.

As it's a ground-up rewrite, I doubt the 'changed files' view on GitHub will be useful, I'd recommend cloning the repo and browsing the code yourself. To run: `python -m SimpleHTTPServer`. When you update the code in src, you have to build the js with browserify: `browserify -o main.js src/main.js`. A nice gulpfile would be nice, preferably with a deploy to `gh-pages` task. I anticipate that this pull request will take some time to be digested and moved, and if there are any changes to be made I'd be happy to accept PRs to this branch prior to merging, particularly for the 'logarithmic spectrum' task, or gulpifying everything. I'm not sure why there's a merge conflict, 

Thumbs, please! @nevosegal @jakubfiala 

Thanks to @JamesMilnerUK and @2xaa for helping! :sparkles:

### CODE REVIEW

**Comment by hughrawlinson** [comment]
To see what this looks like, [check out our surge.sh page](https://meyda.surge.sh/). Note that you have to use `https` because Chrome no longer supports `getUserMedia` on insecure domains other than localhost. IIRC this is why I had to use `python -m SimpleHTTPServer` instead of just using `file:///`.

**Comment by jakubfiala** [comment]
looks good! I'll try to go through the code asap, but at a first glance:

- in Safari, the demo doesn't show, and there is no *get chrome and get a life* message either. It would probably be better if it fell back to using a short pleasant audio file if `getUserMedia` is not supported.

- also there is no indication of what features are shown.

- personally I'd like to be able to see what all features offered in Meyda are doing. Also to load in some test files and see how the viz reacts. This could be added by simply dumping bits of the old demo code into this new one, and rendering in like the top right corner of the viz frame. Also there's the cool lissajous by @hughrawlinson from [meyda-presentation](https://github.com/jakubfiala/meyda-presentation) that would be an interesting addition.

Maybe that ^ is not really important, in which case I don't mind, it's cool as it is.

**Comment by hughrawlinson** [comment]
Agreed on the 'fallback to short pleasant audio' idea. I hear Elvis' early work is coming into the public domain, might be the most recognisable recordings we'd have permissions for?

I'll add a list of what each feature arrow is.

While I was still experiencing major performance issues, I did try to add more features. Unfortunately, there are many that don't really map to this 3d visualisation in any intuitive way. Flux could vary on the z axis, and loudness per bin could be placed in front of the waveform, but I have no idea how we'd do things like kurtosis. Ideas very welcome.

**Comment by jakubfiala** [comment]
I'd say just a simple number output with a dropdown list to choose which feature is being displayed would be great. At least for the features that don't output arrays.

**Comment by jakubfiala** [comment]
And yes, Elvis is perfect, agreed!

**Comment by hughrawlinson** [comment]
Oh, I see what you mean. I was thinking this particular display would be an plain-ish general viewer, with a more complex viewer on another page, letting you watch your own audio files, with a scrubber and maybe looping, along with (of course) checking the audio features. What do you think?

**Comment by jakubfiala** [comment]
That sounds like a good idea! That way people would instantly see what the lib can do on the first page, and if they want to experiment with it, they'd get a separate view for that. 👍

**Comment by 2xAA** [comment]
You may want to integrate MediaStreamTrack's getSources to allow the user to switch between audio sources easily. I was confused as to why nothing seemed to be working when I entered the page, turns out I was set on a digital input source which had nothing routed through it.

Also you may want a notice on the page or a redirect if the user is not on the HTTPS Github pages URL to ensure GetUserMedia works.

Looks good other than those couple of points. Keep up the great work!

**Comment by hughrawlinson** [comment]
@2xAA Good idea, I'll do that. Is there a reasonable way to automatically detect an active source?

Good shout on the redirect too, I'll leave myself a link here so I know how to do it: https://konklone.com/post/github-pages-now-sorta-supports-https-so-use-it

Looks like GitHub still doesn't officially support SSL :frowning: https://github.com/isaacs/github/issues/156

**Comment by jakubfiala** [comment]
@hughrawlinson i remember there was a way to do SSL with the free tier CloudFlare account, not sure if it still works though

**Comment by 2xAA** [comment]
@hughrawlinson Well the default system source will always be selected first which will most likely be a microphone or the most used audio input which the user will be familiar with. As long as you make it clear there's a dropdown selection to the user they'll figure it out I'm sure.

Otherwise you'd probably have to loop through the IDs and detect a peak value over 0 or something which I probably wouldn't recommend doing, seem like too much work when you can just write "Select your audio source to test Meyda" 😄 

Good luck on the SSL thing. I naively thought it was enabled as a full service just by adding 'https'.

**Comment by hughrawlinson** [comment]
Sorry I'm going slow on this, work has been busy. Updates:

* SSL is hard to do for free (I'll have https only though, so no need for a redirect notice)
* Haven't done anything with getSources yet
* Safari breaks because it doesn't support block scope variables yet. Gonna have to use Babel because I'm fairly sure some of the rendering optimizations rely on the improved garbage collection from block scoping. If we don't want to depend on Babel then we may have to manually mess about with scoping by immediately calling anonymous functions, which I'd really rather not get into.
* I have no idea what to do about IE/Edge

**Comment by jakubfiala** [comment]
@hughrawlinson again, I unfortunately can't do this myself atm, but personally I'd *always* use Babel on everything at this stage. Same thing with vendor-prefixed web audio. That way you can make it work on anything that has the API in any form. ES6 is irrelevant in this case, and it will be easy to update the code in a couple of years when the standard will be supported everywhere. Also Babel is already a dependency of the actual meyda package, isn't it?

**Comment by hughrawlinson** [comment]
Can't believe I didn't think of this before, but how about we just host this site on surge? They've got SSL built in, and we can link there from the repository. We could just redirect from GitHub pages. We could put this website in another repository.

Things left to resolve (I will edit this comment if I've missed anything):

- [ ] SSL
- [ ] Source selection
- [ ] Babelify

I'll set SSL to checked and go with the surge SSL site idea, let me know if there are any objections and we can reconsider.

**Comment by skratchdot** [comment]
I've also been seeing more and more about GitLab: https://about.gitlab.com/2016/04/11/tutorial-securing-your-gitlab-pages-with-tls-and-letsencrypt/

Feels kinda dirty linking from Github though.

**Comment by hughrawlinson** [comment]
@skratchdot I'd do that if it has SSL with no effort, but that tutorial looks like there's >0 effort.

**Comment by skratchdot** [comment]
haha. Yeah- there's a lot of steps, plus the whole "refreshing the cert" part w/ let's encrypt.  Not sure if that's automated w/ Gitlab or not.

**Comment by hughrawlinson** [comment]
Found [a googlewhack](https://www.google.se/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=mediatrackstream), didn't even know that was possible anymore.

**Comment by hughrawlinson** [comment]
Also: `MediaStreamTrack.getSources is deprecated. See https://www.chromestatus.com/feature/4765305641369600 for more details.`

TL;DR is: they want us to use a method that is supported in only Chrome and Firefox.

I vote we call crappy browser support and do it later. An argument for: multiple audio input devices is rare, the default being incorrect is rarer, and the chances of users with multiple input devices being able to understand what the problem is and deciding to listen to Elvis instead are high. I could just connect both streams to Meyda for this demo and not hide the audio tag.

I'm also in favour of perfecting later rather than blocking this PR on issues like these, mostly because none of us realistically have more than a couple of hours a month if that to spend on this, and that time is better spent implementing audio feature extractors. That was the whole point, right? 😛 

Babelify update is still in progress.

**Comment by hughrawlinson** [comment]
Ok, this should be ready to commit. `npm run build` and `npm run serve` should work and help with development (after `npm install` of course). We use Babel and Watchify for this.

Seperately, does Safari seriously still require webkit prefixes for Web Audio? God damn it I hate this.

**Comment by hughrawlinson** [comment]
Ok, so now I've really messed up git history :P This is what I get for working on multiple computers. Don't worry, I'll rebase or squash or something.

Current status: Meyda works in Safari again, I just need to do a deploy. This also fixes #134. Unfortunately, SPN is giving us zeroes in Safari even for the audio tag. I'll continue investigation, and squash the commits once I'm done.

**Comment by jakubfiala** [comment]
MediaElementSource doesn't work in Saf. Use a BufferSource.

**Comment by hughrawlinson** [comment]
Ok, we're done! The page [hosted on surge.sh](https://meyda.surge.sh) works. Following a merge, gh-pages will redirect users to meyda.surge.sh. This site uses the copy of meyda that's in master now (following today's updates that are about to be deployed). All is tutti frutti. 👍s please!

**Comment by hughrawlinson** [comment]
@jakubfiala MediaElementSource does work in Safari, but only on `https`. See [here](https://meyda.surge.sh/).

**Comment by nevosegal** [comment]
:+1:

**Comment by jakubfiala** [comment]
damn, that's great, i didn't know it worked at all! awesome, 👍
